### DISCUSSION TITLE
How to hide or anonymize client ip before sending to upstream proxy?

### DISCUSSION THREAD

**Discussion by TheFantasticLoki** [abhinavsingh/proxy.py#1337] [Q&A]
Hello I started setting up a proxy.py instance after failing to get proxybroker's server to work. I've got everything working as best as I can so far and I've made a script to find proxies, check them, then use them as proxy pools in proxy.py and it seems to be working. My problem is that when I do a curl to https://ipinfo.io/json through the proxy.py instance I still get my own ip as the result when I want it to be anonymized. I don't really get what I need to do and chatgpt seems to not be able to help me any further.

Here is my script so far in case anyone is curious:
```
#!/bin/bash

# Function to check if a proxy is working
check_proxy() {
  local ip="$1"
  local port="$2"

  if curl --proxy "$ip:$port" -sSfI --max-time 1 "https://ipinfo.io/json" >/dev/null; then
    echo -e "\033[32mWorking proxy\033[0m: $ip:$port"
    return 0
  else
    echo -e "\033[31mNon-working proxy\033[0m: $ip:$port"
    return 1
  fi
}

# Clear contents of working proxies file and proxy pool
working_proxies_file="./working_proxies.txt"

echo "" > "$working_proxies_file"
unset proxy_pool

# Grab US proxies and save to file
proxybroker find --types HTTP HTTPS --lvl High --countries US --strict -l 25 | awk '/[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+:[0-9]+/{match($0,/[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+:[0-9]+/); print substr($0,RSTART,RLENGTH)}' > ips.txt

while read -r line <&3; do
  ip=$(echo "$line" | cut -d':' -f1)
  port=$(echo "$line" | cut -d':' -f2)

  if check_proxy "$ip" "$port"; then
    echo "$line" >> "$working_proxies_file"
    proxy_pool+=" --proxy-pool $line"
  fi
done 3< "ips.txt"

# Run final command
command="proxy --disable-headers via --port 8899 --plugins proxy.plugin.ProxyPoolPlugin$proxy_pool"
echo $command
eval "$command"
```
It can definately use some work but it works well enough rn. Any help would be appreciated.

**Comment by TheFantasticLoki** [comment]
@abhinavsingh @alexey-pelykh I've done more testing and tried to disable the following headers in order to remove the clients IP before being sent to the upstream proxies: Via, X-Forwarded-For, Remote_Addr, and X-Real-IP. Unfortunately this does not seem to work as I intended it to since upon doing another `curl -v https://ipinfo.io/json --proxy http://localhost:8899` I still get my own IP address. When I was using proxybroker serve I was having issues with it rarely connecting and not showing connections at all but at least I would get a different IP returned 1 out of every 15 or so tries so ik it is at least possible to get the ip to change via a proxy.

In my attempt to resolve this I also attempted to have chatgpt code me a plugin that would modify the X-Forwarded-For header to change the ip to one of the upstream proxies ips rather than removing the ip all together. Unfortunately I haven't been able to test if this plugin would work since I can't seem to get it to load in proxy.py.
Following the unable to load plugins guide in the readme, I tried to specify the path to the .py file of the plugin but got the following error: `TypeError: the 'package' argument is required to perform a relative import for '.workspaces.python.proxy-plugins.changeip.x_forwarded_for_plugin'` So after that I tried to export the path to the plugin file using `export PYTHONPATH=$PYTHONPATH:/workspaces/python/proxy-plugins`
This also failed to work so I then tried moving the x_forwarded_for_plugin.py file to the plugin folder inside proxy.py then manually add in the import into the _init_.py file but then I got the following error: `ModuleNotFoundError: No module named 'x_forwarded_for_plugin'`
I don't really know what is preventing me from loading this plugin so I can test it, but I'll try to do more testing and experimenting in the mean time. Hopefully one of you can help me and HOPEFULLY it's not actually this complicated.

**Comment by abhinavsingh** [comment]
@TheFantasticLoki Removing HTTP header alone will not hide your IP address.   Remember, TCP connections are made between 2-IP addresses.

1) You can take inspiration from this to randomise IP address for outgoing requests https://github.com/abhinavsingh/proxy.py#customnetworkinterface
2) Another option is to use our proxy pool plugin and setup remote proxies in various parts of the world, so that your original IP is not revealed

Best

**Reply by TheFantasticLoki** [reply]
If you had taken the time to look at the script I was using you would already know that I have proxy pools setup already and my ip is still being returned. But I will check out the Custom Network Interface plugin when I have time.
